File: cpp/4_passport_processing/solution.cpp
/* Author: Thomas Mortier
 */
#include <chrono>
#include <string>
#include <fstream>
#include <iostream>
#include <set>
#include <regex>

const std::set<std::string> F {"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"};

bool checkfield(std::string fid, std::string fval)
{
    bool valid {false};
    if (fid.compare("byr")==0)
        valid = (1920<=std::stoi(fval))&&(std::stoi(fval)<=2002);
    else if (fid.compare("iyr")==0)
        valid = (2010<=std::stoi(fval))&&(std::stoi(fval)<=2020);
    else if (fid.compare("eyr")==0)
        valid = (2020<=std::stoi(fval))&&(std::stoi(fval)<=2030);
    else if (fid.compare("hgt")==0)
    {
        if (regex_match(fval,std::regex("^.*(in|cm)$")))
        {
            std::string hgttype {fval.substr(fval.length()-2,2)};
            unsigned int hgtval {static_cast<unsigned int>(std::stoul(fval.substr(0,fval.length()-2)))};
            if (hgttype.compare("cm")==0)
                valid = (150<=hgtval)&&(hgtval<=193);
            else if (hgttype.compare("in")==0)
                valid = (59<=hgtval)&&(hgtval<=76);
        }
    }
    else if (fid.compare("hcl")==0)
        valid = regex_match(fval,std::regex("^#[0-9a-f]{6}$"));
    else if (fid.compare("ecl")==0)
        valid = regex_match(fval,std::regex("^(amb|blu|brn|gry|grn|hzl|oth)$"));
    else if (fid.compare("pid")==0)
        valid = regex_match(fval,std::regex("^[0-9]{9}$"));
    else if (fid.compare("cid")==0)
        valid = true;
    return valid;
}

unsigned int checkpsw(std::string psw)
{
    bool valid {true}, found {false};
    for (auto& f: F)
    {
        unsigned long ind {0};
        found = false;
        while (ind!=std::string::npos)
        {
            std::string idval {psw.substr(ind,psw.length()-ind)};
            auto pospswsp {psw.substr(ind,psw.length()-ind).find(' ')};
            if (pospswsp!=std::string::npos)
                idval = psw.substr(ind,pospswsp);
            std::string fid {idval.substr(0,idval.find(':'))};
            std::string fval {idval.substr(idval.find(':')+1)};
            if (f.compare(fid)==0)
                found = true;
            valid = checkfield(fid, fval);
            if (valid!=true)
                break; 
            ind = (pospswsp==std::string::npos ? std::string::npos : ind+pospswsp+1);
        }
        if ((valid!=true)||(found!=true))
            break;
    }
    return (valid&&found);
}

int main(int argc, char** argv)
{
    auto t1 = std::chrono::high_resolution_clock::now();
    std::string file {static_cast<std::string>(argv[1])};
    std::ifstream in {file};
    std::string line;
    std::string psw {""};
    unsigned long cntr {0};
    try 
    {
        while (std::getline(in, line))
        {
            if (line.compare("")==0)
            {
                cntr += checkpsw(psw.substr(0,psw.size()-1));
                psw = "";
            }
            else
                psw += line + ' ';
        }
        cntr += checkpsw(psw.substr(0,psw.size()-1));
    }
    catch (std::ifstream::failure e)
    {
        std::cerr << "Ooops, something went wrong: " << e.what() << '\n';
        exit(1);
    }
    std::cout << cntr << std::endl;
    auto t2 = std::chrono::high_resolution_clock::now();
    auto time = std::chrono::duration_cast <std::chrono::milliseconds>(t2-t1).count();
    std::cout << "Executed in " << time << " ms\n"; 
    return 0;
}

File: python/4_passport_processing/solution.py
"""
Author: Thomas Mortier
"""
import re
import sys
import time

t0 = time.time()
fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
psw, cntr = "", 0

def checkField(field):
    valid = False
    if field[0] == "byr":
        valid = (1920<=int(field[1])<=2002)
    elif field[0] == "iyr": 
        valid = (2010<=int(field[1])<=2020)
    elif field[0] == "eyr":
        valid = (2020<=int(field[1])<=2030)
    elif field[0] == "hgt":
        if field[1][-2:] == "cm":
            valid = (150<=int(field[1][:-2])<=193)
        elif field[1][-2:] == "in":
            valid = (59<=int(field[1][:-2])<=76)
    elif field[0] == "hcl":
        valid = bool(re.match('^#[0-9a-f]{6}$',field[1]))
    elif field[0] == "ecl":
        valid = bool(re.match('^(amb|blu|brn|gry|grn|hzl|oth)$',field[1]))
    elif field[0] == "pid":
        valid = bool(re.match('^[0-9]{9}$',field[1]))
    elif field[0] == "cid":
        valid = True
    return valid 

def checkPsw(psw):
    global fields
    idvals = [tuple(t.split(":")) for t in psw.split(" ")]
    valid, found = True, False
    for f in fields:
        found = False
        for fid, fval in idvals:
            if f == fid:
                found = True
            valid = checkField((fid,fval))
            if not valid:
                break
        if not valid or not found:
            break
    return valid and found

for line in open(sys.argv[1],'r'):
    line_str = line.strip()
    if line_str=="":
        cntr += checkPsw(psw[:-1])
        psw = ""
    else:
        psw += line.strip() + ' '
else:
    cntr += checkPsw(psw[:-1])
t1 = time.time()
print(cntr)
print("Executed in {0} s".format(t1-t0))
